pkg/fleet-manager/fleet_plugin.go

```diff
@@ -0,0 +1,492 @@
+	if !f.helmReleaseReady(ctx, fleet) {
+		return nil, fmt.Errorf("wait for HelmRelease to be ready")
+	}
+
+	log.V(4).Info("begin to reconcile extra resources for metric plugin")
+	if err := f.reconcileExtraResources(ctx, fleet); err != nil {
+		return nil, err
+	}
+
+	return resources, nil
+}
+
+func (f *FleetManager) deleteMetricPlugin(ctx context.Context, fleet *fleetapi.Fleet) error {
+	// delete HelmRelease
+	fleetLabels := fleetMetricResourceLables(fleet.Name)
+	fleetLabelsSelector := labels.SelectorFromValidatedSet(map[string]string(fleetLabels)).String()
+	err := f.Client.DeleteAllOf(ctx, &hrapiv2b1.HelmRelease{}, client.InNamespace(fleet.Namespace), fleetLabels)
+	if err != nil {
+		return err
+	}
+	// delete HelmRepository
+	err = f.Client.DeleteAllOf(ctx, &sourcev1.HelmRepository{}, client.InNamespace(fleet.Namespace), fleetLabels)
+	if err != nil {
+		return err
+	}
+	// TODO: delete objstore secret form member cluster
+	for _, cluster := range fleet.Spec.Clusters {
+		c := cluster
+		clusterKey := types.NamespacedName{Namespace: fleet.Namespace, Name: c.Name}
+		fleetCluster, err := f.getFleetCluster(ctx, c.Kind, clusterKey)
+		if err != nil {
+			return err
+		}
+
+		kclient, err := f.clientForCluster(fleet.Namespace, fleetCluster)
+		if err != nil {
+			return err
+		}
+
+		if err := kclient.KubeClient().CoreV1().Secrets(MonitoringNamespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: fleetLabelsSelector,
+		}); err != nil {
+			return err
+		}
+	}
+
+	// delete thanos-sidecar-remote service
+	err = f.Client.DeleteAllOf(ctx, &corev1.Service{}, client.InNamespace(fleet.Namespace), fleetLabels)
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (f *FleetManager) helmReleaseReady(ctx context.Context, fleet *fleetapi.Fleet) bool {
+	// Wait for all helm releases to be reconciled
+	hrList := hrapiv2b1.HelmReleaseList{}
+	if err := f.Client.List(ctx, &hrList, client.InNamespace(fleet.Namespace), client.MatchingLabels(fleetMetricResourceLables(fleet.Name))); err != nil {
+		return false
+	}
+
+	for _, tc := range hrList.Items {
+		if !isReady(tc.Status.Conditions) {
+			log.Info("helm release is not ready", "helm release", tc.Name)
+			return false
+		}
+	}
+
+	return true
+}
+
+// reconcileSidecarService reconciles thanos-sidecar-remote service endpoints, which is a headless service
+// TODO: find a better way to collect service endpoints of thanos-sidecar-remote service after all helm releases are reconciled
+func (f *FleetManager) reconcileExtraResources(ctx context.Context, fleet *fleetapi.Fleet) error {
+	endpoints := sets.New[string]()
+	for _, cluster := range fleet.Spec.Clusters {
+		c := cluster
+		clusterKey := types.NamespacedName{Namespace: fleet.Namespace, Name: c.Name}
+		fleetCluster, err := f.getFleetCluster(ctx, c.Kind, clusterKey)
+		if err != nil {
+			return err
+		}
+
+		kclient, err := f.clientForCluster(fleet.Namespace, fleetCluster)
```

Comment:
can we cache and reuse the client? currently in one reconcile we will create multi times